pkg/serverless/daemon/daemon.go

```diff
@@ -337,7 +337,8 @@ func (d *Daemon) TellDaemonRuntimeDone() {
 // WaitForDaemon waits until the daemon has finished handling the current invocation
 func (d *Daemon) WaitForDaemon() {
 	// We always want to wait for any in-progress flush to complete
-	d.FlushWg.Wait()
+	d.FlushLock.Lock()
+	d.FlushLock.Unlock() //nolint:staticcheck
```

Comment:
I am not sure to understand the Lock/Unlock?

Reply:
`Lock` will block until the mutex is unlocked, and then calling `Unlock` right away will prevent `WaitForDaemon` from blocking something else.

Reply:
Yeah, thanks for asking, please check my thinking on this.

I figure, this `WaitForDaemon` method needs to 1) determine if a flush is happening, and 2) if so, wait until the flush is done. We were originally using a wait group, which keeps an internal counter, if the counter is >0 we wait until it becomes 0.

However, Is there any situation we can imagine where the wait group counter would be anything but 1 or 0? If not, then we can replace the wait group with a mutex. A mutex is thread safe whereas the wait group is not.

Reply:
When I touched this bit of code, I seem to recall thinking that multiple flushes could happen simultaneously. By replacing the the WaitGroup with a Mutex, we force flushes to be serial and prevent a situation in which multiple flushes keep the waitgroup positive. I suppose the tradeoff here is that it's possible for a `flush` to be stuck behind `WaitForDaemon`. I remember if that's a problem or not.